### PR TITLE
[mlir][linalg] Add support to pass attributes to the packed ops

### PR DESCRIPTION
**Pull request by yzhang93**
There is a use case that we need to pass some attributes to the packed op (linalg.generic) during packing or packing transpose. With this change, the attributes will be preserved in the packedOp or transposedOp.

### CODE REVIEW

**Comment by yzhang93** [comment]
@MaheshRavishankar Please review this or pull in anyone else who is required to review the patch.

**Review by MaheshRavishankar — APPROVED** [review]
Thanks!

File: mlir/include/mlir/Dialect/Linalg/Utils/Utils.h [hunk 1]
```diff
@@ -369,7 +369,9 @@ struct GenerateLoopNest {
 /// Returns an attribute list that excludes pre-defined attributes.
 template <typename OpTy>
 SmallVector<NamedAttribute> getPrunedAttributeList(OpTy op) {
-  auto elidedAttrs = llvm::to_vector(op.getAttributeNames());
+  // op.getAttributeNames() doesn't work when the op is linalg::LinalgOp.
+  // Instead use the static function to get attribute names.
+  auto elidedAttrs = llvm::to_vector(linalg::GenericOp::getAttributeNames());
```
Comment (hanhanW) [inline mlir/include/mlir/Dialect/Linalg/Utils/Utils.h:374]: But this is a function template. Why can you assume that all the operations have attributes as identical as a GenericOp?

File: mlir/include/mlir/Dialect/Linalg/Utils/Utils.h [hunk 2]
```diff
@@ -369,7 +369,9 @@ struct GenerateLoopNest {
 /// Returns an attribute list that excludes pre-defined attributes.
 template <typename OpTy>
 SmallVector<NamedAttribute> getPrunedAttributeList(OpTy op) {
-  auto elidedAttrs = llvm::to_vector(op.getAttributeNames());
+  // op.getAttributeNames() doesn't work when the op is linalg::LinalgOp.
```
Comment (hanhanW) [inline mlir/include/mlir/Dialect/Linalg/Utils/Utils.h:372]: What do you mean by does not work? Can you elaborate or add a test/example to the PR...? Should we just have a template partial specialization for LinalgOp?

**Review by hanhanW — CHANGES_REQUESTED** [review]

File: mlir/include/mlir/Dialect/Linalg/Utils/Utils.h (hunk 2 above)
Comment (MaheshRavishankar) [reply mlir/include/mlir/Dialect/Linalg/Utils/Utils.h:372]: `LinalgOp` is an interface op... so it doesnt have that method. All `Linalg` ops have the same attributes as a `linalg.generic`. So this works. We could do partial specialization of `LinalgOp`. That would be cleaner. cc @yzhang93

Comment (hanhanW) [reply mlir/include/mlir/Dialect/Linalg/Utils/Utils.h:372]: yeah, that's also my guess, but I'm not sure. Please also add it to the comment.. Adding a partial specialization makes sense to me.

**Review by nicolasvasilache — CHANGES_REQUESTED** [review]
This needs much more polishing: we have support for op name, interfaces etc in the match op, please adopt/adapt/generalize as needed.

Also, no landing without tests.

**Comment by MaheshRavishankar** [comment]
> This needs much more polishing: we have support for op name, interfaces etc in the match op, please adopt/adapt/generalize as needed.
> 
> Also, no landing without tests.

The match op only works for transform dialect (unless I am missing something). This is outside of transform dialect.

**Comment by yzhang93** [comment]
Close this PR because we have a workaround downstream.

**Pull request closed without merging**